baserun/instrumentation/openai.py

```diff
@@ -0,0 +1,184 @@
+        },
+    ]
+
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return _instruments
+
+    @staticmethod
+    def set_request_attributes(span: Span, kwargs: dict[str, Any]):
+        span.set_attribute(SpanAttributes.LLM_VENDOR, OPENAI_VENDOR_NAME)
+        span.set_attribute(SpanAttributes.OPENAI_API_BASE, openai.api_base)
+        span.set_attribute(SpanAttributes.OPENAI_API_TYPE, openai.api_type)
+
+        if "temperature" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_TEMPERATURE, kwargs.get("temperature")
+            )
+
+        if "top_p" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_TOP_P, kwargs.get("top_p"))
+
+        if "frequency_penalty" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_FREQUENCY_PENALTY,
+                kwargs.get("frequency_penalty"),
+            )
+
+        if "presence_penalty" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_PRESENCE_PENALTY,
+                kwargs.get("presence_penalty"),
+            )
+
+        if "functions" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_FUNCTIONS, json.dumps(kwargs.get("functions"))
+            )
+
+        if "function_call" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_FUNCTION_CALL,
+                json.dumps(kwargs.get("function_call")),
+            )
+
+        if "n" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_N, kwargs.get("n"))
+
+        if "stream" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_STREAM, kwargs.get("stream"))
+
+        if stop := kwargs.get("stop"):
+            if isinstance(stop, str):
+                span.set_attribute(SpanAttributes.LLM_CHAT_STOP_SEQUENCES, [stop])
+            else:
+                span.set_attribute(SpanAttributes.LLM_CHAT_STOP_SEQUENCES, stop)
+
+        if "logit_bias" in kwargs:
+            span.set_attribute(
+                SpanAttributes.LLM_LOGIT_BIAS, json.dumps(kwargs.get("logit_bias"))
+            )
+
+        if "logprobs" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_BEST_OF, kwargs.get("logprobs"))
+
+        if "echo" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_ECHO, kwargs.get("echo"))
+
+        if "suffix" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_SUFFIX, kwargs.get("suffix"))
+
+        if "best_of" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_BEST_OF, kwargs.get("best_of"))
+
+        if "user" in kwargs:
+            span.set_attribute(SpanAttributes.LLM_USER, kwargs.get("user"))
+
+        messages = kwargs.get("messages", [])
+        for i, message in enumerate(messages):
+            prefix = f"{SpanAttributes.LLM_PROMPTS}.{i}"
+            span.set_attribute(f"{prefix}.role", message.get("role"))
+
+            if content := message.get("content"):
```

Comment:
I think we need to handle function calls and name properties here as well

Reply:
The `functions` and `function_call` parameters are on the request itself, and not on the message. On the response they're on the message. Very confusing.

Reply:
I don't think that's quite right. The functions / function_call parameters are used to say which functions are available. But if you pass multiple messages you can still provide previous function call response in the full list

Reply:
Look at the typescript type on the server side

Reply:
I believe this comment still stands. OK to follow-up